src/api/trace/tracer.rs

```diff
@@ -131,3 +133,140 @@ impl<S: Tracer> TracerGenerics for S {
         result
     }
 }
+
+/// `SpanBuilder` allows span attributes to be configured before the span
+/// has started.
+///
+/// ```rust
+/// use opentelemetry::{
+///     api::{self, Provider, Tracer},
+///     global,
+/// };
+///
+/// let tracer = global::trace_provider().get_tracer("example-tracer");
+///
+/// // Use span builder to assign span properties
+/// let _span = tracer
+///     .span_builder("example-span-name")
+///     .with_kind(api::SpanKind::Server)
+///     .start(&tracer);
+/// ```
+#[derive(Debug)]
+pub struct SpanBuilder {
+    /// Parent `SpanContext`
+    pub parent_context: Option<api::SpanContext>,
+    /// Span id, useful for integrations with external tracing systems.
+    pub span_id: Option<u64>,
+    /// Span kind
+    pub span_kind: Option<api::SpanKind>,
+    /// Span name
+    pub name: String,
+    /// Span start time
+    pub start_time: Option<SystemTime>,
+    /// Span end time
+    pub end_time: Option<SystemTime>,
+    /// Span attributes
+    pub attributes: Option<Vec<api::KeyValue>>,
+    /// Span Message events
+    pub message_events: Option<Vec<api::Event>>,
+    /// Span Links
+    pub links: Option<Vec<api::Link>>,
+    /// Span status
+    pub status: Option<api::SpanStatus>,
+}
+
+/// SpanBuilder methods
+impl SpanBuilder {
+    /// Create a new span builder from a span name
+    pub fn from_name(name: String) -> Self {
```

Comment:
Could we take a tracer reference as a parameter here? It means we wouldn't need to pass the tracer into the `Start` fn below.

Reply:
Yeah it definitely is not ideal to have to pass a reference to start. Seems like there are a few options:

The expected API would be something like:

```rust
let _span = tracer
    .span_builder("example-span-name")
    .with_kind(api::SpanKind::Server)
    .start();
```

The simple way to accomplish this would be for `SpanBuilder` to have a type parameter:

```rust
impl SpanBuilder<T: api::Tracer> {
    pub fn from_name(name: String, tracer: &T) -> Self {}
    pub fn start(self) -> T::Span {}
   ...
}
```

However that would mean the `Tracer` impl would look like:

```rust
pub trait Tracer {
    fn span_builder(&self) -> SpanBuilder<Self> {}
}
```

And we need `Tracer`s to be trait objects when used in the `global` module so they cannot have `Self` in parameters or return types.

Another option could be that instead people use `SpanBuilder` directly:

```rust
    SpanBuilder::from_name("example-span-name", &tracer).start()
```

But that isn't very ergonomic either.

The current solution seems like a middle ground, `Tracer`s can still be trait objects as their methods do not reference `Self`, but it requires an explicit reference on span builder's `start`.

Definitely open to alternatives if I've missed a better option that allows for the ideal API above. Thoughts?

Reply:
Alternatively `SpanBuilder` could accept an `&dyn api::Tracer` and return a `Box<dyn api::Span>`? Tradeoff would be having an explicitly boxed span as the return type, but that might be preferable to the tracer reference?

Reply:
Hey - sorry for slow reply.

I think using `Box<..>` types for the tracer and span would be okay. Would it be possible to support both your initial and boxed tracer options?

Reply:
Yeah can keep trying. Might be another option that doesn't have these downsides...

Reply:
@MikeGoldsmith I think this is the best we can get currently with two options:

```rust
 // The builder can be used to create a span directly with the tracer
 let _span = tracer.build(SpanBuilder {
     name: "example-span-name".to_string(),
     span_kind: Some(SpanKind::Server),
     ..Default::default()
 });

 // Or used with builder pattern
 let _span = tracer
     .span_builder("example-span-name")
     .with_kind(SpanKind::Server)
     .start(&tracer);
```

The builder has a hard time taking a reference because the types can't be known ahead of time, e.g. `tracer: &dyn Tracer<Span=?>` can't encode that without needing the builder to be generic, which doesn't currently work in the global module. 

A middle ground is having the builder impl `Default` so it can be used as a form of default args as I show above. The upside of the start method taking a reference to the tracer is builders can be passed around without worrying about the tracer lifetime or having to clone tracers for every builder.

Any other ideas on how this API could be nicer? or should we go with this for now

Reply:
I think the second example, where we use the builder pattern is the most readable. Let's go in that direction for now, and we can re-visit in the future if needed - this PR has been hanging around for awhile so lets get it merged :smile:

Reply:
Ok great, the PR currently supports both patterns I pasted there so we should be good to go if you want to review 🎉